src/vulkan-renderer/world/collision_query.cpp

```
@@ -0,0 +1,107 @@
+                                             const glm::vec3 direction) const {
+    glm::vec3 inverse_dir{1 / direction.x, 1 / direction.y, 1 / direction.z};
+    std::int32_t sign[3]{inverse_dir.x < 0, inverse_dir.y < 0, inverse_dir.z < 0};
+
+    float tmin, tmax, tymin, tymax, tzmin, tzmax;
+
+    tmin = (box_bounds[sign[0]].x - position.x) * inverse_dir.x;
+    tmax = (box_bounds[1 - sign[0]].x - position.x) * inverse_dir.x;
+    tymin = (box_bounds[sign[1]].y - position.y) * inverse_dir.y;
+    tymax = (box_bounds[1 - sign[1]].y - position.y) * inverse_dir.y;
+
+    if ((tmin > tymax) || (tymin > tmax)) {
+        return false;
+    }
+    if (tymin > tmin) {
+        tmin = tymin;
+    }
+    if (tymax < tmax) {
+        tmax = tymax;
+    }
+
+    tzmin = (box_bounds[sign[2]].z - position.z) * inverse_dir.z;
+    tzmax = (box_bounds[1 - sign[2]].z - position.z) * inverse_dir.z;
+
+    if ((tmin > tzmax) || (tzmin > tmax)) {
+        return false;
+    }
+
+    return true;
+}
+
+std::optional<RayCubeCollision<Cube>> OctreeCollisionQuery::check_for_collision(const glm::vec3 pos,
+                                                                                const glm::vec3 dir) {
+    // If the cube is empty, a collision with a ray is not possible,
+    // and there are no sub cubes to check for collision either.
+    if (m_cube.type() == Cube::Type::EMPTY) {
+        // No collision found.
+        return std::nullopt;
+    }
+
+    // We need to pass this into glm::intersectRaySphere by reference,
+    // although we are not interested into it.
+    auto intersection_distance{0.0f};
+
+    const float sphere_radius_squared = static_cast<float>(std::pow(m_cube.bounding_box_radius(), 2));
+
+    // First, check if ray collides with bounding sphere.
+    // This is much easier to calculate than a collision with a bounding box.
+    if (glm::intersectRaySphere(pos, dir, m_cube.center(), sphere_radius_squared, intersection_distance)) {
```

Fail early:
if (!glm::intersectRaySphere(pos, dir, m_cube.center(), sphere_radius_squared, intersection_distance)) {
  return std::nullopt;
}

---

Done.